lib/storage/src/rbac/mod.rs

```diff
@@ -1,9 +1,209 @@
-pub mod access;
-pub mod access_level;
-pub mod access_token;
-#[allow(dead_code)]
-pub mod collection_access;
-#[allow(dead_code)]
-pub mod error;
-
 mod ops_checks;
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+
+use segment::json_path::JsonPath;
+use segment::types::ValueVariants;
+use serde::{Deserialize, Serialize};
+
+use crate::content_manager::errors::StorageError;
+
+/// A structure that defines access rights.
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+#[serde(untagged)]
+pub enum Access {
+    /// Cluster-wide access.
+    Cluster(ClusterAccessMode),
+    /// Access to specific collections.
+    Collection(AccessCollection),
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+pub struct AccessCollection(pub Vec<AccessCollectionRule>);
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+pub struct AccessCollectionRule {
+    /// Collection names that are allowed to be accessed
+    pub collections: Vec<String>,
+
+    pub access: CollectionAccessMode,
+
+    /// Payload constraints.
+    /// An object where each key is a JSON path, and each value is JSON value.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub payload: Option<PayloadConstraint>,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Debug)]
+pub enum ClusterAccessMode {
```

Comment:
Added explicit `allows_as_much_as` method instead of deriving `Ord`.

Reply:
Is there a good reason for not using `is_read()` and `is_write()` functions instead? Because this would still be incompatible if we'd add a new variant that does not fit in this linear state of permission.

Reply:
Replaced `allows_as_much_as` with this structure:
 
```rust
pub struct AccessRequrements {
    /// Write access is required.
    pub write: bool,
    /// Manage access is required, implies write access.
    pub manage: bool,
    /// If true, the access should be not limited by a payload restrictions.
    pub whole: bool,
}
```

So it could be checked like this:
```rust
access.check_collection_access(&collection_name, AccessRequrements::new().write().whole())?;
```